src/app/context/wallet/WalletProvider.tsx

```diff
@@ -1,70 +1,205 @@
+}
+
+const CosmosWalletContext = createContext<CosmosWalletContextProps>({
+  bech32Address: "",
+  pubKey: "",
+  connected: false,
+  disconnect: () => {},
+  signArbitrary: async () => {},
 });
 
 export const WalletProvider = ({ children }: PropsWithChildren) => {
-  const [walletProvider, setWalletProvider] = useState<IWalletProvider>();
+  const [btcWalletProvider, setBTCWalletProvider] =
+    useState<IBTCWalletProvider>();
   const [network, setNetwork] = useState<networks.Network>();
   const [publicKeyNoCoord, setPublicKeyNoCoord] = useState("");
   const [address, setAddress] = useState("");
-  const [connectModalOpen, setConnectModalOpen] = useState(false);
+
+  const [cosmosWalletProvider, setCosmosWalletProvider] = useState<any>();
+  const [cosmosBech32Address, setCosmosBech32Address] = useState("");
+  const [cosmosPubKey, setCosmosPubKey] = useState("");
+  const [cosmosChainID, setCosmosChainID] = useState("");
 
   const { showError } = useError();
 
-  const disconnect = useCallback(() => {
-    setWalletProvider(undefined);
+  const tomoModal = useTomoModalControl();
+  const tomoWalletConnect = useTomoWalletConnect();
+
+  const tomowalletState = useTomoWalletState();
+  const providers = useTomoProviders();
+
+  const disconnect = useCallback(async () => {
+    setBTCWalletProvider(undefined);
     setNetwork(undefined);
     setPublicKeyNoCoord("");
     setAddress("");
-  }, []);
+    setCosmosWalletProvider(undefined);
+    setCosmosBech32Address("");
+    setCosmosPubKey("");
+    await tomoWalletConnect.disconnect();
+  }, [tomoWalletConnect]);
+
+  const open = useCallback(async () => {
+    await tomoModal.open("connect");
+  }, [tomoModal]);
+
+  const connectCosmos = useCallback(async () => {
+    if (!providers.cosmosProvider) return;
+
+    try {
+      const chainID = providers.cosmosProvider.getChainId();
+      const cosmosInfo =
+        await providers.cosmosProvider.provider.getKey(chainID);
+
+      const { bech32Address, pubKey } = cosmosInfo;
+      setCosmosWalletProvider(providers.cosmosProvider);
+      setCosmosBech32Address(bech32Address);
+      setCosmosPubKey(Buffer.from(pubKey).toString("hex"));
+      setCosmosChainID(chainID);
+    } catch (error: any) {
+      showError({
+        error: {
+          message: error.message,
+          errorState: ErrorState.WALLET,
+        },
+        retryAction: connectCosmos,
+      });
+    }
+  }, [providers.cosmosProvider, showError]);
 
-  const open = useCallback(() => {
-    setConnectModalOpen(true);
-  }, []);
+  const cosmosContextValue = useMemo(
+    () => ({
+      bech32Address: cosmosBech32Address,
+      pubKey: cosmosPubKey,
+      connected: Boolean(cosmosWalletProvider),
+      disconnect,
+      open,
+      signArbitrary: async (message: string) =>
+        await cosmosWalletProvider.provider.signArbitrary(
+          cosmosChainID,
+          cosmosBech32Address,
+          message, // This might be Buffer.from(message)
+        ),
```

Comment:
```suggestion
      signArbitrary: (message: string) =>
        cosmosWalletProvider.provider.signArbitrary(
          cosmosChainID,
          cosmosBech32Address,
          message, // This might be Buffer.from(message)
        ),
```

Reply:
@totraev `signArbitrary` is definitely `async` operation since it waits for the user modal on Keplr side

Reply:
Removed here also